server/src/main/java/org/opensearch/gateway/ClusterStateUpdaters.java

```diff
@@ -120,7 +121,19 @@ static ClusterState updateRoutingTable(final ClusterState state) {
         // initialize all index routing tables as empty
         final RoutingTable.Builder routingTableBuilder = RoutingTable.builder(state.routingTable());
         for (final IndexMetadata cursor : state.metadata().indices().values()) {
-            routingTableBuilder.addAsRecovery(cursor);
+            // Whether IndexMetadata is recovered from local disk or remote it doesn't matter to us at this point.
+            // We are only concerted about index data recovery here. Which is why we only check for remote store enabled and not for remote
+            // cluster state enabled.
+            if (cursor.getSettings().getAsBoolean(IndexMetadata.SETTING_REMOTE_STORE_ENABLED, false) == false
+                || state.routingTable()
+                    .shardsMatchingPredicateCount(
+                        shardRouting -> shardRouting.primary()
+                            // We need to ensure atleast one of the primaries is being recovered from remote.
+                            // This ensures we have gone through the RemoteStoreRestoreService and routing table is updated
+                            && shardRouting.recoverySource() instanceof RecoverySource.RemoteStoreRecoverySource
```

Comment:
it is checking for all shards in the routingTable and not specific to current index. What is the reasoning for this logic though you are running this logic in the for loop

Reply:
this is a miss in refactoring. `shardsMatchingPredicateCount` is supposed to run for the specific index only. will update this

Reply:
update the code and added a UT to cover this scenario as well.